Implements nearbyint

khurd21 opened the pull request:
<!-- Please target the "dev" branch when creating a pull request to ensure you are using the latest version of ccmath -->
Implements `ccm::nearbyint` based on [cppreference](https://en.cppreference.com/w/cpp/numeric/math/nearbyint)

Test cases test `ccm::nearbyint<T>` for int, double, long double, and float values. Additionally, `ccm::nearbyintl` and `ccm::nearbyintf` are tested.

Rinzii commented on include/ccmath/math/nearest/nearbyint.hpp:
```diff
@@ -8,7 +8,64 @@
+		T int_part	= ccm::trunc(num);
+		T frac_part = num - int_part;
+		if (frac_part == T(0.0)) { return int_part; }
+		if (frac_part > T(0.5) || (frac_part == T(0.5) && static_cast<long long>(int_part) % 2 != 0)) { return int_part + T(1); }
```
Casting to long long here introduces some risks due to the implementation-defined behavior across different platforms. While long long is guaranteed to be at least 64 bits, the cast may lead to precision loss or unexpected behavior, especially when dealing with long double. This issue might not manifest on all platforms—on some, the cast may work as expected—but it’s safer to avoid relying on this behavior. Instead, we should consider using our helper class `ccmath/tree/main/include/ccmath/internal/support/fp/fp_bits.hpp` to directly manipulate the floating-point bits. This approach will ensure more consistent and reliable handling, particularly for long double, and mitigate the risks associated with platform-specific integer representations. The class also handles a lot of the tricky bits in bit manipulation also. ^^

khurd21 replied:
Thanks! I will look into replacing this.

Rinzii commented on include/ccmath/math/nearest/nearbyint.hpp:
```diff
@@ -8,7 +8,64 @@
 namespace ccm
 {
+	/**
+	 * @brief The nearest integer value to num, according to the rounding mode FE_TONEAREST, is returned.
```
Are we able to support all rounding modes? This should be pretty do-able with `fp_bits' and the `fp` folder in our support directory should have all the necessary tools to enable us to properly support all rounding modes. ^^

khurd21 replied:
I can surely look into it! I will look into the fp_bits in the project, thanks for bringing that up. I was initially unsure if this was something that was possible as the standard library implementation seemed very runtime reliant, this was why I opted to just implement what the default rounding mode seemed to be.

Rinzii commented on include/ccmath/math/nearest/nearbyint.hpp:
```diff
@@ -8,7 +8,64 @@
+	template <class Integer, std::enable_if_t<std::is_integral_v<Integer>, bool> = true>
+	constexpr double nearbyint(Integer num) noexcept
+	{
+		return static_cast<double>(num);
```
This is fine for our needs and you don't need to change this.

Rinzii commented on include/ccmath/math/nearest/nearbyint.hpp:
```diff
@@ -8,7 +8,64 @@
+	 */
+	constexpr long double nearbyintl(long double num)
+	{
+		return nearbyint(num);
```
Please apply the full qualified namespace of the function. This is mostly to be a proactive measure to avoid possible issues with compilers getting confused. Should not happen, but we have no clue what the end users environment will look like so better safe than getting a bug report later.

khurd21 replied:
Clarification question. Do you mean change the function to be something like:

```cpp
constexpr long double nearbyintl(long double num) {
    return ccm::nearbyint(num);
 }
 ```

Rinzii replied:
Yes that is what I want. Makes things more explicit. Generally when calling functions that could theoretically have a collision I'm generally just explicit with it. ^^

Rinzii reviewed (changes requested):
Overall this is a great start! I am being a bit critical since this is a main function implementation and I generally hold all implementations in high regard so sorry if I'm asking a lot! Still great work and I am excited to see your updated changes. ^^

Rinzii commented on test/nearest/nearbyint_test.cpp:
```diff
@@ -0,0 +1,305 @@
+// Disabling test case ccm::nearbyintl if run on clang linux.
+#ifdef __clang__
+	#ifdef __linux__
+		#define CLANG_LINUX
```
This is actually a pretty new issue for me. Looking over clang everything is working as intended. It also appears to really just be an issue with fp precision when dealing with long double with bit_cast. GCC and clang also appear to have different schools of thought on how the size issue should be handled with bit_cast. GCC just does not care at assumes long double is the correct size even if there is a possibility they could be different. Where Clang appears to more so treat them as incorrect and prevents the operation from happening. Currently I don't know of any easy way to resolve this sadly, but I'll look into it later when I have more time and see if I can find a better work around. ^^

khurd21 commented:
@Rinzii 

Here are some of the changes I added:
- `ccm::` namespace to functions called within the file.
- Additional rounding mode support
- Leveraged the fp internal support files.

I looked into the support/fp section and noticed you had two files: `rounding_mode.hpp` and `directional_rounding_utils.hpp`. It looks like all of this behavior is already contained in there, so I replaced my code with that. I wanted to ask if this was the accepted approach. It looks like it is; it seems to already be used in some capacity within the `ccm::pow` impl file minus the `get_rounding_mode` free function. It also already seems to have all of the rounding behaviors implemented using the fp bits you suggested I use. I was initially going to go down the manual implementation route, but this seemed to be a better alternative.

Look forward to the re-review. No problem on the "picky-ness", I am still learning the library!

Rinzii commented:
> @Rinzii
> 
> Here are some of the changes I added:
> 
>     * `ccm::` namespace to functions called within the file.
> 
>     * Additional rounding mode support
> 
>     * Leveraged the fp internal support files.
> 
> 
> I looked into the support/fp section and noticed you had two files: `rounding_mode.hpp` and `directional_rounding_utils.hpp`. It looks like all of this behavior is already contained in there, so I replaced my code with that. I wanted to ask if this was the accepted approach. It looks like it is; it seems to already be used in some capacity within the `ccm::pow` impl file minus the `get_rounding_mode` free function. It also already seems to have all of the rounding behaviors implemented using the fp bits you suggested I use. I was initially going to go down the manual implementation route, but this seemed to be a better alternative.
> 
> Look forward to the re-review. No problem on the "picky-ness", I am still learning the library!

Ah yes I forgot I had implemented that helper class before to solve some internal problems for the exact same reason that I told you earlier lol. 

Also so see some examples of some of the best implementations of the library that in my opinion are complete I'd check out this file:

https://github.com/Rinzii/ccmath/blob/6b76871afcecde3c478cb075866347bdbbe5a93a/include/ccmath/internal/generic/functions/power/sqrt_gen.hpp#L147

Also if you have any questions or need directions on possible functionality already implemented feel free to ask them inside of the relevant issues! ^^

Rinzii reviewed (approved):
lgtm!
